pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,353 @@
+				err = ds.clearPods()
+				if err != nil {
+					err = util.Errorf("Unable to clear pods from intent tree: %v", err)
+					select {
+					case errCh <- err:
+					case <-quitCh:
+					}
+				}
+				return
+
+			case <-nodesChangedCh:
+				ds.logger.NoFields().Infof("Received node update signal")
+				if ds.Disabled {
+					continue
+				}
+				err := ds.removePods()
+				if err != nil {
+					err = util.Errorf("Unable to remove pods from intent tree: %v", err)
+					continue
+				}
+				err = ds.addPods()
+				if err != nil {
+					err = util.Errorf("Unable to add pods to intent tree: %v", err)
+					continue
+				}
+
+			case <-quitCh:
+				return
+			}
+		}
+	}()
+
+	return errCh
+}
+
+func (ds *daemonSet) meetDesires() error {
+	ds.logger.NoFields().Infof("Handling DS update: disabled %v", ds.Disabled)
+
+	// Do nothing if disabled
+	if ds.Disabled {
+		return nil
+	}
+
+	return nil
+}
+
+// addPods schedules pods for all unscheduled nodes selected by ds.nodeSelector
+func (ds *daemonSet) addPods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	eligible, err := ds.scheduler.EligibleNodes(ds.Manifest, ds.NodeSelector)
+	if err != nil {
+		return util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to schedule on and then sort them
+	// for deterministic ordering
+	toScheduleSorted := types.NewNodeSet(eligible...).Difference(types.NewNodeSet(currentNodes...)).ListNodes()
+	ds.logger.NoFields().Infof("Need to schedule %d nodes", len(toScheduleSorted))
+
+	for _, node := range toScheduleSorted {
+		err := ds.schedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+// removePods unschedules pods for all scheduled nodes not selected
+// by ds.nodeSelector
+func (ds *daemonSet) removePods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	eligible, err := ds.scheduler.EligibleNodes(ds.Manifest, ds.NodeSelector)
+	if err != nil {
+		return util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to unschedule on and then sort them
+	// for deterministic ordering
+	toUnscheduleSorted := types.NewNodeSet(currentNodes...).Difference(types.NewNodeSet(eligible...)).ListNodes()
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+// clearPods unschedules pods for all the nodes that have been scheduled by
+// this daemon set by using CurrentNodes()
+// This should only be used when a daemon set is deleted
```

Comment:
the comment here suggests `clearPods()` should only be called when deleting a daemon set, but above it is used when disabled